app/src/main/java/com/github/gotify/login/AdvancedDialog.java

```diff
@@ -41,32 +46,56 @@ AdvancedDialog onClickRemoveCaCertificate(Runnable onClickRemoveCaCertificate) {
         return this;
     }
 
-    AdvancedDialog show(boolean disableSSL, @Nullable String selectedCertificate) {
+    AdvancedDialog onClickSelectClientCertificate(Runnable onClickSelectClientCertificate) {
+        this.onClickSelectClientCertificate = onClickSelectClientCertificate;
+        return this;
+    }
+
+    AdvancedDialog onClickRemoveClientCertificate(Runnable onClickRemoveClientCertificate) {
+        this.onClickRemoveClientCertificate = onClickRemoveClientCertificate;
+        return this;
+    }
 
+    AdvancedDialog show(boolean disableSSL, @Nullable String selectedCaCertificate, @Nullable String selectedClientCertificate, String password, Settings settings) {
         View dialogView =
                 LayoutInflater.from(context).inflate(R.layout.advanced_settings_dialog, null);
         holder = new ViewHolder(dialogView);
         holder.disableSSL.setChecked(disableSSL);
         holder.disableSSL.setOnCheckedChangeListener(onCheckedChangeListener);
+        holder.editClientCertPass.setText(password);
 
-        if (selectedCertificate == null) {
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            holder.editClientCertPass.setVisibility(View.GONE);
+            holder.selectedClientCertificate.setVisibility(View.GONE);
+            holder.toggleClientCert.setVisibility(View.GONE);
+        }
+
+        if (selectedCaCertificate == null) {
             showSelectCACertificate();
         } else {
-            showRemoveCACertificate(selectedCertificate);
+            showRemoveCACertificate(selectedCaCertificate);
+        }
+
+        if (selectedClientCertificate == null) {
+            showSelectClientCertificate();
+        } else {
+            showRemoveClientCertificate(selectedClientCertificate);
         }
 
         new AlertDialog.Builder(context)
                 .setView(dialogView)
                 .setTitle(R.string.advanced_settings)
-                .setPositiveButton(context.getString(R.string.done), (ignored, ignored2) -> {})
+                .setPositiveButton(context.getString(R.string.done), (ignored, ignored2) -> {
+                    settings.clientCertPass(holder.editClientCertPass.getText().toString());
```

Comment:
Settings should only be set inside `com.github.gotify.login.LoginActivity#onCreatedClient`

Reply:
This is my first android code, so help me here. The other buttons on the advanced screen start file selection activities that end up being handled by an `onActivityResult` in `LoginActivity.java`. I was not sure how to get the value of the client cert password except through a lambda that is run after the done button is clicked. Do you have another way? I do see settings being modified in other functions besides `onCreatedClient`.

Reply:
I'm running out of things to try in creating the self-signed certificate. I've tried a couple of different things and the LOG on the login page says `Hostname .... not verified`. I've tried selecting my CA root and the server cert and I can see that it is trying to validate the CA cert. This is all on the master branch by the way.

Reply:
Ok this is what worked for me:

I was getting all my cert files confused and was trying to use the wrong file for the client cert. Also, by using the Caddy auto-cert I am getting around some apparent issue with how I was trying to manually create the CA and server cert (a problem for another day). In my case, the client cert and client CA are both separately generated and totally unrelated to the server cert (Caddy autogen cert or otherwise).

### Commands to create the client certificate:
```
openssl req -x509 -newkey rsa:2048 -keyout gotify_client_cert.key -out gotify_client_cert.crt -days 3000
openssl req -new -key gotify_client_cert.key -out gotify_client_cert.csr
openssl x509 -req -days 3000 -in gotify_client_cert.csr -signkey gotify_client_cert.key -out gotify_client_cert-CA.crt
cat gotify_client_cert.crt gotify_client_cert.key > gotify_client_cert.pem
openssl pkcs12 -export -out gotify_client_cert.p12 -inkey gotify_client_cert.key -in gotify_client_cert.pem
```

### Caddyfile:
```
(require_client_cert) {
    tls {
        client_auth {
            mode require_and_verify
            trusted_ca_cert_file /client_certs/gotify_client_cert-CA.crt
            trusted_leaf_cert_file /client_certs/gotify_client_cert.crt
        }
    }
}

* *.home.lan {
    import require_client_cert
    tls internal
    reverse_proxy http://gotify:80
}
```

### Docker compose file:
```
version: '3.3'
services:
  caddy:
    image: caddy:2.4.6-alpine
    ports:
      - "3080:80"
      - "3443:443"
    volumes:
      - ./Caddyfile:/etc/caddy/Caddyfile
      - ./.ssh:/client_certs
  gotify:
    image: gotify/server:2.1.4
volumes:
  caddy_data1:
```

Reply:
I'll have a look at this on the weekend.